lib/coffeescript/command.js

```diff
@@ -605,54 +605,50 @@
+        }
         process.exit(1);
       }
-      // We’re giving Babel only a string, not a filename or path to a file, so
-      // it doesn’t know where to search to find a `.babelrc` file or a `babel`
-      // key in a `package.json`. So if `opts.transpile` is an object, use that
-      // as Babel’s options; otherwise figure out what the options should be.
       if (typeof opts.transpile !== 'object') {
-        // Find the options based on the path to the file being compiled.
-        cantFindOptions = function() {
-          console.error('To use the transpile option, there must be a .babelrc file\n(or a package.json file with a "babel" key) in the path of the file\nto be compiled, or in the path of the current working directory.\nIf you are compiling a string via the Node API, the transpile option\nmust be an object with the options to pass to Babel.\nSee http://coffeescript.org/#transpilation');
-          return process.exit(1);
-        };
-        checkPath = filename ? path.dirname(filename) : base ? base : typeof process !== "undefined" && process !== null ? process.cwd() : cantFindOptions();
-        while (true) {
-          try {
-            opts.transpile = JSON.parse(fs.readFileSync(path.join(checkPath, '.babelrc'), 'utf-8'));
-            break;
-          } catch (error) {
-            try {
-              packageJson = JSON.parse(fs.readFileSync(path.join(checkPath, 'package.json'), 'utf-8'));
-              if (packageJson.babel != null) {
-                opts.transpile = packageJson.babel;
-                break;
-              }
-            } catch (error) {}
-          }
-          if (checkPath === path.dirname(checkPath)) { // We’ve reached the root.
-            cantFindOptions();
-            break;
-          } else {
-            checkPath = path.dirname(checkPath);
-          }
-        }
+        opts.transpile = {};
       }
       // Pass a reference to Babel into the compiler, so that the transpile option
       // is available for the CLI. We need to do this so that tools like Webpack
       // can `require('coffeescript')` and build correctly, without trying to
       // require Babel.
       opts.transpile.transpile = CoffeeScript.transpile;
+      // Babel searches for its options (a `.babelrc` file, a `.babelrc.js` file,
+      // a `package.json` file with a `babel` key, etc.) relative to the path
+      // given to it in its `filename` option. Make sure we have a path to pass
+      // along.
+      if (!opts.transpile.filename) {
+        if (filename) {
+          opts.transpile.filename = filename;
+        } else {
+          opts.transpile.filename = base || process.cwd();
+          if (opts.transpile.filename.endsWith(path.sep)) {
+            opts.transpile.filename += '.';
```

Comment:
For the same of making the error messages clearer, I wonder if it's be better to have this do `<stdio>.coffee` `<fake>.coffee` or something? Not sure what cases come up where the filename isn't known.

I guess relatedly, _should_ Babel use the `.babelrc` if it doesn't know a filename? Babel's CLI itself for instance does not take `.babelrc` files when code is passed in via `stdin`, though we've talked about adding an opt-in flag for that behavior. I don't have strong feelings how you do it though, since it's probably fine as long as it's documented for users.

Reply:
The only cases I can think of where `filename` would be undefined are when the `coffee` command is called with input via `stdin` or via `--eval`, e.g.:

```bash
./bin/coffee --transpile --eval 'import path from "path"
console.log path.sep
throw new Error()'
/
Error
    at Object.<anonymous> (<anonymous>:1:54)
    at Module._compile (module.js:573:30)
    at Object.CoffeeScript.run (~/Sites/coffeescript/lib/coffeescript/index.js:61:23)
```

And the CoffeeScript compiler is already rewriting the stack trace in that scenario, putting `<anonymous>` in place of the filename.

That covers exceptions, but not the Babel compiler itself throwing an error:

```bash
./bin/coffee --transpile --eval '`import path`'
SyntaxError: ~/Sites/coffeescript/.: 'import' and 'export' may only appear at the top level (2:2)
  1 | (function() {
> 2 |   import path;
    |   ^
  3 |
  4 |
  5 | }).call(this);
    at Parser.pp$5.raise (~/Sites/coffeescript/node_modules/babel-core/node_modules/babylon/lib/index.js:4454:13)
    at Parser.pp$1.parseStatement (~/Sites/coffeescript/node_modules/babel-core/node_modules/babylon/lib/index.js:1877:16)
    at Parser.pp$1.parseBlockBody (~/Sites/coffeescript/node_modules/babel-core/node_modules/babylon/lib/index.js:2268:21)
```

So the `filename` appears next to the `SyntaxError` here, but I think `/.` here is fine. I think it might make a bit more sense than `~/Sites/coffeescript/./<anonymous>`, which would cause some head scratching since that path wouldn’t resolve. Also I’m still nervous about so much reliance on `filename`, which is essentially undocumented; what if a future change to `filename` means that it needs to be a resolvable path? Then suddenly passing it `~/Sites/coffeescript/./<anonymous>` breaks `--transpile`.

> _should_ Babel use the `.babelrc` if it doesn’t know a filename?

Yes, to handle the `--eval` or `stdin` cases. Is there a reason we shouldn’t allow `--transpile` to work in those scenarios? I’d rather Babel find its options the usual way based on the current folder for these cases, than have someone pass in JSON as a command-line argument.

Reply:
> So the filename appears next to the SyntaxError here, but I think /. here is fine.

Sounds good, just figured I'd mention it.

> Also I’m still nervous about so much reliance on filename, which is essentially undocumented

Babel's documentation just isn't great but that doesn't mean we're anxious to break our users. We do our absolute best to only introduce breaking changes on major versions. There's zero chance we'd go from treating `filename` as a maybe-existing file to a guaranteed-existing file without a major version bump.

> I’d rather Babel find its options the usual way based on the current folder for these cases

I think it's your call. I don't feel strongly. I'll say, the reason in my mind that having the filename is important for config resolution is for cases like this:

```
project/
  src/
    index.js
    .babelrc
```
given

```
babel src/index.js
// vs
babel < src/index.js
```
because the first will find the `.babelrc` and the second one won't.

For Babel, we just skip config resolution entirely unless we know the actual location to search from.

Reply:
@loganfsmyth Thanks for all these notes. I think the `--eval` and `stdin` use cases are edge enough that I’d rather pass the current folder as the filename so that transpilation works for these options, rather than force someone to save a `test.coffee` file or similar. Most likely someone compiling code inline via the CLI is just testing things, and this saves them some hassle. We should know (and pass along) the filename for all common use cases, as far as I’m aware, whether they’re compiling a single file or a folder tree.

Do you mind adding a comment in the Babel source code near where `filename` is parsed that upstream tools are expecting this functionality out of it? Perhaps with a link to this PR or to https://github.com/jashkenas/coffeescript/issues/4713 😄 Any little bit helps.

Reply:
I don't think adding a comment isn't gonna make it any clearer. It's a core piece of the way Babel processes configs. The fact that it exists in the first place by definition means that upstream things rely on it heavily and changing it would break them.